addons/html_editor/static/src/others/dynamic_placeholder/dynamic_placeholder.js

```diff
@@ -0,0 +1,92 @@
+
+export class DynamicPlaceholderPlugin extends Plugin {
+    static name = "dynamicplaceholder";
+    static dependencies = ["selection", "history", "dom"];
+    static resources = (p) => ({
+        powerboxCategory: { id: "marketing_tools", name: _t("Marketing Tools"), sequence: 60 },
+        powerboxCommands: {
+            name: _t("Dynamic Placeholder"),
+            description: _t("Insert personalized content"),
+            category: "marketing_tools",
+            fontawesome: "fa-magic",
+            action(dispatch) {
+                dispatch("OPEN_DYNAMIC_PLACEHOLDER");
+            },
+        },
+    });
+    setup() {
+        if (!this.config.dynamicPlaceholder || this.config.dynamicPlaceholderModelReferenceField) {
+            throw new Error("Invalid Dynamic Placeholder Config");
+        }
+        this.dynamicPlaceholder = this.config.dynamicPlaceholder;
+        this.dynamicPlaceholderModelReferenceField =
+            this.config.dynamicPlaceholderModelReferenceField;
+    }
+
+    handleCommand(command) {
+        switch (command) {
+            case "OPEN_DYNAMIC_PLACEHOLDER": {
+                this.openDialog();
+
+                this.eventListenerHash = this.addDomListener(
+                    this.editable,
+                    "beforeinput",
+                    this.onBeforeInput.bind(this),
+                    true
+                );
+                break;
+            }
+        }
+    }
+
+    openDialog() {
+        this.selectionToRestore = this.shared.getEditableSelection();
+
+        this.dynamicPlaceholder.updateModel(this.dynamicPlaceholderModelReferenceField);
+        this.dynamicPlaceholder.open({
+            validateCallback: this.onDynamicPlaceholderValidate.bind(this),
+            closeCallback: this.onDynamicPlaceholderClose.bind(this),
+            positionCallback: this.positionDynamicPlaceholder.bind(this),
+        });
+    }
+    onBeforeInput(ev) {
+        if (ev.inputType === "insertParagraph") {
+            ev.preventDefault();
+            ev.stopImmediatePropagation();
+        }
+    }
+
+    onDynamicPlaceholderValidate(chain, defaultValue) {
+        if (chain) {
+            // Ensure the focus is in the editable document
+            // before inserting the <t> element.
+            let dynamicPlaceholder = "object." + chain;
+            dynamicPlaceholder +=
+                defaultValue && defaultValue !== "" ? ` or '''${defaultValue}'''` : "";
+            const t = document.createElement("T");
+            t.setAttribute("t-out", dynamicPlaceholder);
+
+            this.shared.domInsert(t);
+            // todo : should focus() in case of iframe ?
+            this.dispatch("NORMALIZE", { node: t.parentElement });
+            this.dispatch("ADD_STEP");
+        }
+    }
+    onDynamicPlaceholderClose() {
+        this.shared.setSelection(this.selectionToRestore);
+        setTimeout(() => this.removeDomListener(this.eventListenerHash), 50);
```

Comment:
Why do we need a timeout to remove the dom listener?
But more importantly why do we need to remove the listener? Can we just set a flag (so that we don't need to introduce removeDomListener)?
`this.dynamicPlaceholderOpened = false`
And in `onBeforeInput`:
`if (this.dynamicPlaceholderOpened && ev.inputType === "insertParagraph") {`